### ISSUE TITLE
Focusrite Scarlett 4th gen drivers not recognized by Rocksmith windows 11

### ISSUE THREAD

**Issue by Armedorrr** [mdias/rs_asio#405]
Only getting Focusrite Thunderbolt ASIO driver to show up in log
works fine for older drivers but older drivers dont support 4th gen interfaces.
Newer drivers work fine for my laptop with windows 10 but not for my desktop with windows 11

RS_ASIO-log

0.000 [INFO]   - Wrapper DLL loaded (v0.7.0)
0.000 [INFO]  PatchOriginalCode
0.028 [INFO]  image crc32: 0xd1b38fcb
0.160 [INFO]  Patching CoCreateInstance
0.160 [INFO]  Patch_CallAbsoluteIndirectAddress - num locations: 2
0.160 [INFO]  Patching call at 00C02DBD
0.160 [INFO]  Patching call at 00E75278
0.160 [INFO]  Patching PortAudio MarshalStreamComPointers
0.160 [INFO]  Patch_CallRelativeAddress - num locations: 1
0.160 [INFO]  Patching call at 00E7483F
0.160 [INFO]  Patching PortAudio UnmarshalStreamComPointers
0.160 [INFO]  Patch_CallRelativeAddress - num locations: 1
0.160 [INFO]  Patching call at 00E748F4
0.160 [INFO]  Patching Two Guitar Tones Connected Message Box (num locations: 1)
0.160 [INFO]  Patching bytes at 007C0C5F
0.162 [INFO]  Patched_CoCreateInstance called: IID_IMMDeviceEnumerator
0.162 [INFO]  DebugDeviceEnum::GetDefaultAudioEndpoint - dataFlow: eRender - role: eMultimedia
0.162 [INFO]  RSAggregatorDeviceEnum::UpdateAvailableDevices
0.162 [INFO]  GetWineAsioInfo - Looking for wineasio.dll... 
0.163 [INFO]    Failed to load wineasio.dll or file not found
0.163 [INFO]  AsioHelpers::FindDrivers
0.163 [INFO]    Focusrite Thunderbolt ASIO
0.163 [INFO]    Realtek ASIO
0.164 [INFO]  RSAsioDeviceEnum::UpdateAvailableDevices - output requesting ASIO driver: Focusrite USB ASIO
0.164 [ERROR]  RSAsioDeviceEnum::UpdateAvailableDevices - failed.
0.164 [INFO]  RSAsioDeviceEnum::UpdateAvailableDevices - input[0] requesting ASIO driver: Focusrite USB ASIO
0.164 [ERROR]  RSAsioDeviceEnum::UpdateAvailableDevices - failed.
0.164 [INFO]  DebugDeviceEnum::UpdateAvailableDevices - 0 render devices, 0 capture devices
0.164 [INFO]    hr: 80070490
0.164 [INFO]    *ppEndpoint: 00000000
0.164 [INFO]  DebugDeviceEnum::GetDefaultAudioEndpoint - dataFlow: eCapture - role: eMultimedia
0.164 [INFO]    hr: 80070490
0.164 [INFO]    *ppEndpoint: 00000000
0.164 [INFO]  DebugDeviceEnum::EnumAudioEndpoints - dataFlow: eAll - dwStateMask: 1
0.164 [INFO]    hr: S_OK
0.164 [INFO]    *ppDevices: 0FD4F6E0
1.685 [INFO]  Patched_CoCreateInstance called: IID_IMMDeviceEnumerator
1.685 [INFO]  DebugDeviceEnum::GetDefaultAudioEndpoint - dataFlow: eRender - role: eMultimedia
1.685 [INFO]  RSAggregatorDeviceEnum::UpdateAvailableDevices
1.685 [INFO]  AsioHelpers::FindDrivers
1.685 [INFO]    Focusrite Thunderbolt ASIO
1.685 [INFO]    Realtek ASIO
1.685 [INFO]  RSAsioDeviceEnum::UpdateAvailableDevices - output requesting ASIO driver: 
1.685 [ERROR]  RSAsioDeviceEnum::UpdateAvailableDevices - failed.
1.685 [INFO]  RSAsioDeviceEnum::UpdateAvailableDevices - input[0] requesting ASIO driver: 
1.685 [ERROR]  RSAsioDeviceEnum::UpdateAvailableDevices - failed.
1.685 [INFO]  DebugDeviceEnum::UpdateAvailableDevices - 0 render devices, 0 capture devices
1.685 [INFO]    hr: 80070490
1.685 [INFO]    *ppEndpoint: 00000000
1.685 [INFO]  DebugDeviceEnum::GetDefaultAudioEndpoint - dataFlow: eCapture - role: eMultimedia
1.685 [INFO]    hr: 80070490
1.685 [INFO]    *ppEndpoint: 00000000
1.685 [INFO]  DebugDeviceEnum::EnumAudioEndpoints - dataFlow: eAll - dwStateMask: 1
1.685 [INFO]    hr: S_OK
1.685 [INFO]    *ppDevices: 0FD4F800
1.690 [INFO]  DebugDeviceEnum::RegisterEndpointNotificationCallback
5.670 [INFO]  DebugDeviceEnum::UnregisterEndpointNotificationCallback
5.802 [INFO]   - Wrapper DLL unloaded

**Comment by Armedorrr** [comment]
Pls get scarlett 4th Gen drivers to work with rs asio
i can only get older drivers to work but the 4th Gen interface does not work with older 3rd gen drivers

**Comment by Armedorrr** [comment]
connected thunderbolt 3 cable from interface to pc 
still not working


pls fix this issue or im going to kms

**Comment by Armedorrr** [comment]
i have resorted to using No Cable Mod since my drivers cant be detected. PLS FIX SOON!

**Comment by Koros1691** [comment]
I just bought a Scarlett 2i2 4th Gen and can confirm things are working just fine for me without anything special in windows 11.

If it helps anyone, my current versions:
RS ASIO 0.7.0
Windows 11 Pro Version 23H2 (OS Build 22631.2861)
Focusrite Control 2 App Version 1.129.0.16711
Focusrite Control 2 Driver Version 4.119.13.15
Scarlett 2i2 4th Gen Firmware Version 2128.0

**Comment by mdias** [comment]
Thank you for the confirmation @Koros1691. I've added it to the interface list 👍

**Comment by songless-bird** [comment]
Also works with Scarlett 16i16 4th Gen, new drivers (Focusrite Control 2) on windows 11. Same rsasio config as with old focusrite drivers. Had to set my buffers a little larger though.

**Comment by mdias** [comment]
Thanks. Added 👍

**Comment by Pranti-cpu** [comment]
Hello, I have a Scarlett Solo 4th Gen, the drivers are detected using the Scarlett Solo 3rd Gen Config File but the sound is not coming out of the Output and not being calibrated in Rocksmith, how do I resolve this ?

**Comment by mdias** [comment]
@Pranti-cpu where do you have your speakers/headphones connected?
And can you please post your config file and log file?

**Comment by vsqla** [comment]
> Also works with Scarlett 16i16 4th Gen, new drivers (Focusrite Control 2) on windows 11. Same rsasio config as with old focusrite drivers. Had to set my buffers a little larger though.

can you share config?

**Comment by songless-bird** [comment]
> > Also works with Scarlett 16i16 4th Gen, new drivers (Focusrite Control 2) on windows 11. Same rsasio config as with old focusrite drivers. Had to set my buffers a little larger though.
> 
> can you share config?

Not sure if I tried using both inputs with the 4th gen yet, but Input 1 works. Sample rate in Focusrite Settings App is 48000 and Buffer Size 48. Clock Source internal. In Windows Audio Settings, output is set to 16bit, 48000 Hz. Input is 2 channel, 24 bit, 48000 Hz right now, but i am quite sure i set it to 16bit several times and it switches back by itself.

RS_ASIO.ini
```
[Config]
EnableWasapi=0
EnableAsio=1

[Asio]
; available buffer size modes:
;    driver - respect buffer size setting set in the driver
;    host   - use a buffer size as close as possible as that requested by the host application
;    custom - use the buffer size specified in CustomBufferSize field
;BufferSizeMode=driver
BufferSizeMode=custom
CustomBufferSize=48

[Asio.Output]
Driver=Focusrite USB ASIO
BaseChannel=0
AltBaseChannel=1
EnableSoftwareEndpointVolumeControl=1
EnableSoftwareMasterVolumeControl=1
SoftwareMasterVolumePercent=100

[Asio.Input.0]
Driver=Focusrite USB ASIO
Channel=0
EnableSoftwareEndpointVolumeControl=1
EnableSoftwareMasterVolumeControl=1
SoftwareMasterVolumePercent=100

;[Asio.Input.1]
;Driver=Focusrite USB ASIO
;Channel=1
;EnableSoftwareEndpointVolumeControl=1
;EnableSoftwareMasterVolumeControl=1
;SoftwareMasterVolumePercent=100
```

**Comment by josipglavas** [comment]
> [@Pranti-cpu](https://github.com/Pranti-cpu) where do you have your speakers/headphones connected? And can you please post your config file and log file?

Got the same issue here. My headphones are connected straight to the scarlett 4th gen, but the game is still not hearing the guitar to be calibrated. 

this is my config: 

## Config file

**RS_ASIO.ini**

```ini
[Config]
EnableWasapiOutputs=0
EnableWasapiInputs=0
EnableAsio=1

[Asio]
BufferSizeMode=driver
CustomBufferSize=48

[Asio.Output]
Driver=Focusrite USB ASIO
BaseChannel=0
EnableSoftwareEndpointVolumeControl=1
EnableSoftwareMasterVolumeControl=1
SoftwareMasterVolumePercent=100

[Asio.Input.0]
Driver=Focusrite USB ASIO
Channel=1
EnableSoftwareEndpointVolumeControl=1
EnableSoftwareMasterVolumeControl=1
SoftwareMasterVolumePercent=100

[Asio.Input.1]
Driver=Focusrite USB ASIO
Channel=2
EnableSoftwareEndpointVolumeControl=1
EnableSoftwareMasterVolumeControl=1
SoftwareMasterVolumePercent=100
```

this is the log from the RS_ASIO-log.txt

0.286 [INFO]  Patched_CoCreateInstance called: IID_IMMDeviceEnumerator
0.287 [INFO]  DebugDeviceEnum::GetDefaultAudioEndpoint - dataFlow: eRender - role: eMultimedia
0.287 [INFO]  RSAggregatorDeviceEnum::UpdateAvailableDevices
0.287 [INFO]  GetWineAsioInfo - Looking for "wineasio32.dll"...  Not found.
0.289 [INFO]  GetWineAsioInfo - Looking for "wineasio.dll"...  Not found.
0.291 [INFO]  AsioHelpers::FindDrivers
0.291 [INFO]    Focusrite Thunderbolt ASIO
0.291 [INFO]    Focusrite USB ASIO
0.291 [INFO]  RSAsioDeviceEnum::UpdateAvailableDevices - output requesting ASIO driver: Focusrite USB ASIO
0.292 [INFO]  Creating AsioSharedHost - dll: c:\program files\focusrite\drivers\focusriteusbasio32.dll
0.301 [INFO]    ASIO input channels info:
0.301 [INFO]      0 - active: 0, channel: 0, group: 0, isInput: 1, type: ASIOSTInt32LSB, name: Input 1
0.301 [INFO]      1 - active: 0, channel: 1, group: 0, isInput: 1, type: ASIOSTInt32LSB, name: Input 2
0.301 [INFO]      2 - active: 0, channel: 2, group: 0, isInput: 1, type: ASIOSTInt32LSB, name: Loopback 1
0.301 [INFO]      3 - active: 0, channel: 3, group: 0, isInput: 1, type: ASIOSTInt32LSB, name: Loopback 2
0.301 [INFO]    ASIO output channels info:
0.301 [INFO]      0 - active: 0, channel: 0, group: 0, isInput: 0, type: ASIOSTInt32LSB, name: Output 1
0.301 [INFO]      1 - active: 0, channel: 1, group: 0, isInput: 0, type: ASIOSTInt32LSB, name: Output 2
0.301 [INFO]  RSAsioDeviceEnum::UpdateAvailableDevices - OK
0.301 [INFO]  RSAsioDeviceEnum::UpdateAvailableDevices - input[0] requesting ASIO driver: Focusrite USB ASIO
0.301 [INFO]  RSAsioDeviceEnum::UpdateAvailableDevices - OK
0.301 [INFO]  RSAsioDeviceEnum::UpdateAvailableDevices - input[1] requesting ASIO driver: Focusrite USB ASIO
0.301 [INFO]  RSAsioDeviceEnum::UpdateAvailableDevices - OK
0.301 [INFO]  DebugDeviceEnum::UpdateAvailableDevices - 1 render devices, 2 capture devices
0.301 [INFO]    hr: S_OK
0.301 [INFO]    *ppEndpoint: 0CCB5750
0.301 [INFO]  DebugDeviceEnum::GetDefaultAudioEndpoint - dataFlow: eCapture - role: eMultimedia
0.301 [INFO]    hr: S_OK
0.301 [INFO]    *ppEndpoint: 0CCB55A0
0.301 [INFO]  DebugDeviceEnum::EnumAudioEndpoints - dataFlow: eAll - dwStateMask: 1
0.301 [INFO]    hr: S_OK
0.301 [INFO]    *ppDevices: 114A67C8
0.301 [INFO]  {ASIO Out} DebugWrapperDevice::OpenPropertyStore - stgmAccess: 0
0.301 [INFO]  {ASIO Out} DebugWrapperDevicePropertyStore::GetCount
0.301 [INFO]    *cProps: 2
0.301 [INFO]  {ASIO Out} DebugWrapperDevice::QueryInterface - riid: IID_IMMEndpoint
0.301 [INFO]  {ASIO Out} DebugWrapperEndpoint::GetDataFlow
0.301 [INFO]    *pDataFlow: eRender

Driver version for scarlett is 4.124.3.5 which i understand is 64bit. 

Sample rate in the focusrite program is set to 48 kHz. 

Game detects the input as RTC but no souind coming in / cannot calibrate

**Comment by vsqla** [comment]
> > > Also works with Scarlett 16i16 4th Gen, new drivers (Focusrite Control 2) on windows 11. Same rsasio config as with old focusrite drivers. Had to set my buffers a little larger though.
> > 
> > 
> > can you share config?
> 
> Not sure if I tried using both inputs with the 4th gen yet, but Input 1 works. Sample rate in Focusrite Settings App is 48000 and Buffer Size 48. Clock Source internal. In Windows Audio Settings, output is set to 16bit, 48000 Hz. Input is 2 channel, 24 bit, 48000 Hz right now, but i am quite sure i set it to 16bit several times and it switches back by itself.
> 
> RS_ASIO.ini
> 
> ```
> [Config]
> EnableWasapi=0
> EnableAsio=1
> 
> [Asio]
> ; available buffer size modes:
> ;    driver - respect buffer size setting set in the driver
> ;    host   - use a buffer size as close as possible as that requested by the host application
> ;    custom - use the buffer size specified in CustomBufferSize field
> ;BufferSizeMode=driver
> BufferSizeMode=custom
> CustomBufferSize=48
> 
> [Asio.Output]
> Driver=Focusrite USB ASIO
> BaseChannel=0
> AltBaseChannel=1
> EnableSoftwareEndpointVolumeControl=1
> EnableSoftwareMasterVolumeControl=1
> SoftwareMasterVolumePercent=100
> 
> [Asio.Input.0]
> Driver=Focusrite USB ASIO
> Channel=0
> EnableSoftwareEndpointVolumeControl=1
> EnableSoftwareMasterVolumeControl=1
> SoftwareMasterVolumePercent=100
> 
> ;[Asio.Input.1]
> ;Driver=Focusrite USB ASIO
> ;Channel=1
> ;EnableSoftwareEndpointVolumeControl=1
> ;EnableSoftwareMasterVolumeControl=1
> ;SoftwareMasterVolumePercent=100
> ```

Thx. There was no "Focusrite USB ASIO" driver, only thunderbird one. Solved this by removing actual driver and installing old focusrite control from 3rd gen. Now i have "Focusrite USB ASIO" and It works like a charm.

**Comment by mdias** [comment]
@josipglavas you cut your log very short, there is not enough information to see if if it's streaming audio or not.

However, please don't use that `[Asio.Input.1]` with `Channel=2` on your interface, as that channel appears to be a loopback channel which would result in a horrible feedback loop.

**Comment by MikeBlackSpain** [comment]
I just got the Focusrite Scarlett 2i2 4th gen - moved on from the Audient ID 4 (great device) which worked no problem with Rocksmith. I copied the .ini above which has **CustomBufferSize=48** but this caused crackling and distortion, 
After changing the **CustomBufferSize=192** (also in the Focusrite device settings) the distortion is gone and I hear no delay or latency. It gives the computer more time to breathe avoiding any glitches.

I also activated the mic [Asio.Input.1] + [Asio.Input.Mic] (I don't know if I need both) which works great unlike the Audient.
If you want this, you must switch the mic on in the settings Rocksmith game - if not, delete or comment the lines out.

--------------------------------------------

[Config]
EnableWasapi=0
EnableAsio=1

[Asio]
; available buffer size modes:
;    driver - respect buffer size setting set in the driver
;    host   - use a buffer size as close as possible as that requested by the host application
;    custom - use the buffer size specified in CustomBufferSize field
BufferSizeMode=custom
CustomBufferSize=192

[Asio.Output]
Driver=Focusrite USB ASIO
BaseChannel=0
AltBaseChannel=1
EnableSoftwareEndpointVolumeControl=1
EnableSoftwareMasterVolumeControl=1
SoftwareMasterVolumePercent=100

[Asio.Input.0]
Driver=Focusrite USB ASIO
Channel=0
EnableSoftwareEndpointVolumeControl=1
EnableSoftwareMasterVolumeControl=1
SoftwareMasterVolumePercent=100

[Asio.Input.1]
Driver=Focusrite USB ASIO
Channel=1
EnableSoftwareEndpointVolumeControl=1
EnableSoftwareMasterVolumeControl=1
SoftwareMasterVolumePercent=100

[Asio.Input.Mic]
Driver= Focusrite USB ASIO
Channel=1
`EnableSoftwareEndpointVolumeControl=1`
EnableSoftwareMasterVolumeControl=1
SoftwareMasterVolumePercent=100

**Comment by ivasquez1999** [comment]
I've got a new Focusrite Scarlett 4th gen with latest firmware on 28/11/2025 driver is 4.143.0.261, firmware is 2.0.2408.0

and the only thing that helped to overcome the crackling is to set like [MikeBlackSpain] here above
BufferSizeMode=custom
CustomBufferSize=192

**Comment by P-Toby** [comment]
I also have a Scarlett 4th gen 2i2 (Firmware 2.0.2417.0, Driver 4.143.0.261) and when I set the customer buffer size parameter in the rs_asio.ini to 48 it immediately worked without any crackling. I changed no other settings beyond just filling out the usual "Focusrite USB ASIO" to use the correct driver.

I have a Ryzen 9 7900 (on W10) so maybe that is a factor? (Assuming you guys that have issues with the lower buffer sizes have older/weaker CPUs?)
